README.md
# proxy-socks

> **TL;DR** follow these [instructions](#instructions) to start building a flexible residential back-connect proxy network on any platform running [Electron](https://electronjs.org/).

# about

<img align="left" src="https://raw.githubusercontent.com/ab77/proxy-socks/master/assets/app-icon/png/48.png">  This [Electron](https://electronjs.org/) desktop app forwards `socksPort` and `squidPort` to `proxyRemote` IP on the remote server over [SSH](https://github.com/mscdex/ssh2) and in reverse, forwards a random port on a remote [Linux box](#server-config) to a [SOCKS](https://github.com/mscdex/socksv5) proxy running locally.

Clients connecting to `squidPort` or `socksPort` on the machine running the app, will be tunneled to `proxyRemote` on remote server. Similarly, remote SOCKS connections will be proxied out the local WAN interface of the machine running the app. It is up to you to handle incoming client connections on the remote server using HAProxy, Squid, etc.


# instructions

<p align="left"><a href="https://m.do.co/c/937b01397c94" target="_blank"><img src="https://raw.githubusercontent.com/ab77/netflix-proxy/master/static/digitalocean.png" width="300"></a></p>

* deploy an Ubuntu or Debian box to be the proxy concentrator and note its public IP (e.g. on [DigitalOcean](https://m.do.co/c/937b01397c94))


## client config
* create SSH key locally and copy to the the proxy concentrator

        mkdir -p ~/.proxy-socks\
          && echo -e 'y\n' | ssh-keygen -f ~/.proxy-socks/id_rsa\
          && ssh-copy-id -i ~/.proxy-socks/id_rsa.pub root@{proxy-concentrator}

* create `config.json`, with the hostname or IP address of the proxy concentrator and adjust ports if required

```
cat << EOF > ~/.proxy-socks/config.json
{
    "username": "tunnel",
    "host": "{proxy-concentrator}",
    "port": 22,
    "privateKey": "${HOME}/.proxy-socks/id_rsa",
    "squidPort": 3128,
    "socksPort": 1080,
    "proxyRemote": "172.20.0.10"
}
EOF
```


## server config
* on the proxy concentrator create `tunnel` user, set a password and authorize keys

        useradd -m tunnel -s /bin/bash\
          && mkdir -p /home/tunnel/.ssh\
          && cat ~/.ssh/authorized_keys > /home/tunnel/.ssh/authorized_keys

* create splash script and set permissions

```
cat << EOF > /home/tunnel/splash.sh
#!/bin/sh
echo '+--------------------------+'
echo '|                          |'
echo '|  Nothing to see here...  |'
echo '|                          |'
echo '+--------------------------+'
EOF

chmod +x /home/tunnel/splash.sh
chown tunnel:tunnel -hR /home/tunnel
```

* update sshd config and restart the service

```
cat << EOF >> /etc/ssh/sshd_config

Match User tunnel
  ForceCommand /home/tunnel/splash.sh
EOF

service ssh restart
```


## installing

The Windows binary is not signed and modern Chrome/Windows will try to prevent it from running. If this is important to you, you can get a signing certificate from one of the [vendors](https://www.sslshopper.com/microsoft-authenticode-certificates.html) and build a signed release.

|OS|release|
|---|---|
|Windows|[latest](https://github.com/ab77/proxy-socks/releases/download/v1.0.5/proxy-socks-setup-1.0.5.exe), [1.0.4](https://github.com/ab77/proxy-socks/releases/download/v1.0.4/proxy-socks-setup-1.0.4.exe), [1.0.3](https://github.com/ab77/proxy-socks/releases/download/v1.0.3/proxy-socks-setup-1.0.3.exe), [1.0.2](https://github.com/ab77/proxy-socks/releases/download/v1.0.2/proxy-socks-setup-1.0.2.exe), [1.0.1](https://github.com/ab77/proxy-socks/releases/download/v1.0.1/proxy-socks-setup-1.0.1.exe), [1.0.0](https://github.com/ab77/proxy-socks/releases/download/v1.0.0/proxy-socks-setup-1.0.0.exe)|
|Linux (AppImage)|[latest](https://github.com/ab77/proxy-socks/releases/download/v1.0.5/proxy-socks-1.0.5-x86_64.AppImage), [1.0.4](https://github.com/ab77/proxy-socks/releases/download/v1.0.4/proxy-socks-1.0.4-x86_64.AppImage), [1.0.3](https://github.com/ab77/proxy-socks/releases/download/v1.0.3/proxy-socks-1.0.3-x86_64.AppImage), [1.0.2](https://github.com/ab77/proxy-socks/releases/download/v1.0.2/proxy-socks-1.0.2-x86_64.AppImage), [1.0.1](https://github.com/ab77/proxy-socks/releases/download/v1.0.1/proxy-socks-1.0.1-x86_64.AppImage), [1.0.0](https://github.com/ab77/proxy-socks/releases/download/v1.0.0/proxy-socks-1.0.0-x86_64.AppImage)|
|Linux (Snap)|[latest](https://github.com/ab77/proxy-socks/releases/download/v1.0.5/proxy-socks_1.0.5_amd64.snap), [1.0.4](https://github.com/ab77/proxy-socks/releases/download/v1.0.4/proxy-socks_1.0.4_amd64.snap), [1.0.3](https://github.com/ab77/proxy-socks/releases/download/v1.0.3/proxy-socks_1.0.3_amd64.snap), [1.0.2](https://github.com/ab77/proxy-socks/releases/download/v1.0.2/proxy-socks_1.0.2_amd64.snap), [1.0.1](https://github.com/ab77/proxy-socks/releases/download/v1.0.1/proxy-socks_1.0.1_amd64.snap), [1.0.0](https://github.com/ab77/proxy-socks/releases/download/v1.0.0/proxy-socks_1.0.0_amd64.snap)|
|Mac OS X|[latest](https://github.com/ab77/proxy-socks/releases/download/v1.0.5/proxy-socks-1.0.5.dmg), [1.0.4](https://github.com/ab77/proxy-socks/releases/download/v1.0.4/proxy-socks-1.0.4.dmg), [1.0.3](https://github.com/ab77/proxy-socks/releases/download/v1.0.3/proxy-socks-1.0.3.dmg), [1.0.2](https://github.com/ab77/proxy-socks/releases/download/v1.0.2/proxy-socks-1.0.2.dmg), [1.0.1](https://github.com/ab77/proxy-socks/releases/download/v1.0.1/proxy-socks-1.0.1.dmg), [1.0.0](https://github.com/ab77/proxy-socks/releases/download/v1.0.0/proxy-socks-1.0.0.dmg)|

* launch the app and note the forwarded port number

<img align="center" src="https://raw.githubusercontent.com/ab77/proxy-socks/master/extra/proxy-socks.png" width="400">

> <img align="left" src="https://raw.githubusercontent.com/ab77/proxy-socks/master/assets/app-icon/png/16.png">  The app will attempt to insert itself into the appropriate start-up chain to start automatically with the operating system and appear minimised to tray.

* test connectivity to the proxy from the proxy concentrator

        # netstat -a -n -p | grep LISTEN | grep 127.0.0.1
        tcp        0      0 127.0.0.1:{zzz}         0.0.0.0:* ...

        root@ubuntu:# curl -4 ifconfig.co
        {xxx}

        root@ubuntu:# curl -4 --socks5 localhost:{zzz} ifconfig.co
        {yyy}

> you should see a random TCP port and two different public IPs if everything is working correctly


# IPv6

> the server is listening on both `AF_INET` and `AF_INET6` sockets

        # netstat -a -n -p | grep LISTEN | grep ::1
        tcp6       0      0 ::1:{zzz}               :::* ...

        root@ubuntu:# curl -6 ifconfig.co
        {xxx}

        root@ubuntu:# curl -6 --socks5 localhost:{zzz} ifconfig.co
        {yyy}


# next steps
Every new installation of the app, will attempt to make a connection to the remote server and forward a random port to the local proxy. These proxies can then be exposed on the public interface of the server using [HAProxy](http://www.haproxy.org/), [OpenVPN](https://openvpn.net/) or a combination of tools.

An [example](https://raw.githubusercontent.com/ab77/proxy-socks/master/extra/update-haproxy) shell script can be used to automatically generate HAProxy configuration for all connected proxies.

> Once the ports are exposed, ensure appropriate ACLs are set.

<hr>
<p align="center">&copy; 2018 <a href="https://anton.belodedenko.me/belodetek/">belodetek</a></p>
<p align="center"><a href="http://anton.belodedenko.me/"><img src="https://avatars2.githubusercontent.com/u/2033996?v=3&s=50"></a></p>


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.0.2
## (2022-11-25)

* Integrate Flowzone CI [belodetek]


renderer.js
const electron = require('electron')
const ipc = electron.ipcRenderer
const startServerBtn = document.getElementById('start-server-btn')
const stopServerBtn = document.getElementById('stop-server-btn')
const startClientBtn = document.getElementById('start-client-btn')
const stopClientBtn = document.getElementById('stop-client-btn')
const connectInput = document.getElementById('connect')
const keyInput = document.getElementById('key')
const passphraseInput = document.getElementById('passphrase')

startServerBtn.disabled = false
stopServerBtn.disabled = true
startClientBtn.disabled = false
stopClientBtn.disabled = true

ipc.on('starting-server', function (event, message) {
  startServerBtn.disabled = true
  document.getElementById('server-info').innerHTML = message
})

ipc.on('started-server', function (event, message) {
  document.getElementById('server-info').innerHTML = message
  if (message.toLowerCase().startsWith('proxy tunnel')) {
    stopServerBtn.disabled = false
  } else {
    startServerBtn.disabled = false
    stopServerBtn.disabled = true
  }
})

ipc.on('stopping-server', function (event, message) {
  stopServerBtn.disabled = true
  document.getElementById('server-info').innerHTML = message
})

ipc.on('stopped-server', function (event, message) {
  document.getElementById('server-info').innerHTML = message
  startServerBtn.disabled = false
})

ipc.on('starting-client', function (event, message) {
  startClientBtn.disabled = true
  document.getElementById('client-info').innerHTML = message
})

ipc.on('started-client', function (event, message) {
  document.getElementById('client-info').innerHTML = message
  if (! message.toLowerCase().startsWith('error')) {
    stopClientBtn.disabled = false
  } else {
    startClientBtn.disabled = false
    stopClientBtn.disabled = true
  }
})

ipc.on('stopping-client', function (event, message) {
  stopClientBtn.disabled = true
  document.getElementById('client-info').innerHTML = message
})

ipc.on('stopped-client', function (event, message) {
  document.getElementById('client-info').innerHTML = message
  startClientBtn.disabled = false
})

ipc.on('connect-set', function (event, message) {
  connectInput.value = message
  ipc.send('connect-input', connectInput.value)
})

ipc.on('key-set', function (event, message) {
  keyInput.value = message
  ipc.send('key-input', keyInput.value)
  ipc.send('start-client')
})

startServerBtn.addEventListener('click', function () {
  ipc.send('start-server')
})

stopServerBtn.addEventListener('click', function () {
  ipc.send('stop-server')
})

startClientBtn.addEventListener('click', function () {
  ipc.send('start-client')
})

stopClientBtn.addEventListener('click', function () {
  ipc.send('stop-client')
})

connectInput.addEventListener('change', function() {
  ipc.send('connect-input', connectInput.value)
})

keyInput.addEventListener('change', function() {
  ipc.send('key-input', keyInput.value)
})

passphraseInput.addEventListener('change', function() {
  ipc.send('passphrase-input', passphraseInput.value)
})

ipc.send('start-server')


main.js
require('string.format')

const net = require('net')
const assert = require('assert')
const fs = require('fs')
const path = require('path')
const url = require('url')
const spawn = require('child_process').spawn
const AutoLaunch = require('auto-launch')
const electron = require('electron')
const app = electron.app
const Menu = electron.Menu
const Tray = electron.Tray
const ipc = electron.ipcMain
const BrowserWindow = electron.BrowserWindow
const Client = require('ssh2').Client
const socks = require('socksv5')
const homedir = require('os').homedir()
const debug = /--debug/.test(process.argv[2])
const default_ssh_port = 22
const default_proxy_remote = '172.20.0.10'    // e.g. Docker container running Squid, etc.
const default_socks_port = 1080
const default_squid_port = 3128

let server
let client
let conn
let socks_proxy
let squid_proxy
let mainWindow
let trayIcon
let trayMenu
let mainMenu
let appName = require('./package.json').name
let resourcesPath = path.resolve(__dirname)
let extraResources = path.join(resourcesPath, 'extra')
if (resourcesPath.endsWith('app.asar')) {
  extraResources = path.resolve(resourcesPath, '..', 'extra')
}

let iShouldQuit = app.makeSingleInstance( () => {
  if (mainWindow) {
    if (mainWindow.isMinimized())
      mainWindow.restore()
    }
    if (! mainWindow.isVisible()) {
      mainWindow.show()
    }
    mainWindow.focus()
    return true
})
if(iShouldQuit) {
  app.isQuiting = true
  app.quit()
  return
}

let dir = path.join(homedir, '.' + appName)
if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir)
}
let configPath = path.resolve(dir)
let configFile = path.join(configPath, 'config.json')

let proxy_remote
try {
  proxy_remote = require(configFile).proxyRemote
  assert.notStrictEqual(proxy_remote, undefined)
} catch (err) {
  console.log(err)
  proxy_remote = default_proxy_remote
}
console.log(proxy_remote)

let squid_port
try {
  squid_port = require(configFile).squidPort
  assert.notStrictEqual(squid_port, undefined)
} catch (err) {
  console.log(err)
  squid_port = default_squid_port
}
console.log(squid_port)

let socks_port
try {
  socks_port = require(configFile).socksPort
  assert.notStrictEqual(socks_port, undefined)
} catch (err) {
  console.log(err)
  socks_port = default_socks_port
}
console.log(socks_port)

let privateKeyPath
try {
  privateKeyPath = require(configFile).privateKey
} catch (err) {
  console.log(err)
}
console.log(privateKeyPath)

let ssh_config = {}
try {
  ssh_config.username = require(configFile).username
  assert.notStrictEqual(ssh_config.username, undefined)
} catch (err) {
  console.log(err)
  ssh_config.username = 'tunnel'
}
try {
  ssh_config.host = require(configFile).host
  assert.notStrictEqual(ssh_config.host, undefined)
} catch (err) {
  console.log(err)
  ssh_config.host = 'localhost'
}
try {
  ssh_config.port = require(configFile).port
  assert.notStrictEqual(ssh_config.port, undefined)
} catch (err) {
  console.log(err)
  ssh_config.port = default_ssh_port
}
console.log(ssh_config)

if (resourcesPath.endsWith('app.asar')) {
  let AutoLauncher = new AutoLaunch({
    name: appName
  })
  AutoLauncher.isEnabled().then(function(isEnabled) {
    if(!isEnabled) {
      AutoLauncher.enable()
    }
  }).catch(function(err){
    console.log(err)
  })
}

function createWindow () {
  let windowOptions = {
    width: 450,
    height: 450,
    title: app.getName(),
    webPreferences: {
      pageVisibility: true,
      backgroundThrottling: false,
    }
  }
  if (process.platform === 'linux') {
    windowOptions.icon = path.resolve(
      path.normalize(
        path.join(__dirname, '/assets/app-icon/png/512.png')
      )
    )
  }

  mainWindow = new BrowserWindow(windowOptions)
  mainWindow.loadURL(url.format({
    pathname: path.join(__dirname, '_index.html'),
    protocol: 'file:',
    slashes: true
  }))

  if (debug) {
    mainWindow.webContents.openDevTools()
    mainWindow.show()
    require('devtron').install()
  } else {
    mainWindow.hide()
  }

  mainWindow.webContents.once('dom-ready', () => {
    try {
      if (fs.existsSync(path.normalize(privateKeyPath))) {
        mainWindow.webContents.send('key-set', privateKeyPath)
      }
    } catch (err) {
      console.log(err)
    }
    ssh_url = ssh_config.username + '@' + ssh_config.host + ':' + ssh_config.port
    mainWindow.webContents.send('connect-set', ssh_url)
  })

  mainWindow.once('ready-to-show', () => {
    mainWindow.hide()
  })

  mainWindow.on('restore', (event) => {
    mainWindow.setSkipTaskbar(false)
    event.preventDefault()
    mainWindow.restore()
    mainWindow.focus()
  })

  mainWindow.on('minimize', (event) => {
    mainWindow.setSkipTaskbar(true)
    event.preventDefault()
    mainWindow.minimize()
  })

  mainWindow.on('close', (event) => {
    if (!app.isQuiting) {
        event.preventDefault()
        mainWindow.hide()
    }
    return false
  })
}

function save_config() {
    let app_config = {
      username: ssh_config.username,
      host: ssh_config.host,
      port: ssh_config.port,
      privateKey: privateKeyPath,
      squidPort: squid_port,
      socksPort: socks_port,
      proxyRemote: proxy_remote
    }
    if (!fs.existsSync(configPath)) {
      fs.mkdirSync(configPath)
    }
    fs.writeFileSync(configFile, JSON.stringify(app_config, null, 4), 'utf8')
}

function forward_port(event, rport, lport) {
  let net = require('net')
  conn = new Client()
  try {
    conn.on('ready', function() {
      conn.forwardIn('::', rport, function(err) {
        if (err) {
          console.log(err)
          event.sender.send('stopped-client', err.message)
          client = undefined
        }
        event.sender.send('started-client', `Forwarding port ${rport} on remote to localhost:${lport}`)
        client = true
        save_config()
      })
    }).on('tcp connection', function(info, accept, deny) {
      let stream = accept()
      stream.pause()
      stream.on('error', function(err) {
        console.log(err)
      })
      let socket = net.connect(lport, '::', function () {
        stream.pipe(socket)
        socket.pipe(stream)
        stream.resume()
      })
      socket.on('error', function(err) {
        console.log(err)
      })
    }).on('error', function(err) {
      console.log(err)
      event.sender.send('stopped-client', err.message)
    }).connect(ssh_config)
  } catch (err) {
    console.log(err)
    event.sender.send('stopped-client', err.message)
  }
}

app.once('ready', () => {
  require('update-electron-app')()
  createWindow()
  trayIcon = new Tray(path.resolve(
    path.normalize(
      path.join(__dirname, '/assets/app-icon/png/16.png')
    )
  ))
  trayMenu = Menu.buildFromTemplate([
  {
    label: 'Show',
    click: () => {
      mainWindow.show()
      mainWindow.focus()
    }
  },
  {
    label: 'Quit',
    click: () => {
      app.isQuiting = true
      app.quit()
    }
  }
  ])
  trayIcon.setContextMenu(trayMenu)
  trayIcon.setToolTip(app.getName())
})

ipc.on('connect-input', (event, message) => {
  try {
    ssh_config.username = message.split(':')[0].split('@')[0]
    ssh_config.host = message.split(':')[0].split('@')[1]
    if (message.split(':')[1] !== undefined) {
      ssh_config.port = message.split(':')[1]
    } else {
      ssh_config.port = default_ssh_port
    }
  } catch (err) {
    console.log(err)
    event.sender.send('stopped-client', err.message)
  }
})

ipc.on('key-input', (event, message) => {
  try {
    privateKeyPath = message
    ssh_config.privateKey = fs.readFileSync(path.normalize(message))
  } catch (err) {
    privateKeyPath = undefined
    ssh_config.privateKey = undefined
    console.log(err)
    event.sender.send('stopped-client', err.message)
  }
})

ipc.on('passphrase-input', (event, message) => {
  if (message) {
    ssh_config.passphrase = message
  }
})

// SOCKS proxy (inbound) listing on random local port
ipc.on('start-client', (event) => {
  if (ssh_config.username && ssh_config.host && ssh_config.port && ssh_config.privateKey) {
    if (client === undefined ) {
      event.sender.send('starting-client', 'Starting remote...')
      if (server === undefined ) {
		server = socks.createServer( (info, accept, deny) => {
		  accept()
		}).listen(0, '::').useAuth(socks.auth.None())
	  }
	  server.on('listening', function() {
	    let lport = server.address().port
 		forward_port(event, 0, lport)
	  })
    } else {
      event.sender.send('started-client', 'Already running')
    }
  } else {
    event.sender.send('stopped-client', 'Not ready')
  }
})

ipc.on('stop-client', (event) => {
  if (server !== undefined) {
    server.close()
    server = undefined
  }
  if (client !== undefined) {
    event.sender.send('stopping-client', 'Stopping remote...')
    conn.end()
    conn = undefined
    client = undefined
    event.sender.send('stopped-client', 'Remote stopped')
  } else {
    event.sender.send('stopped-client', 'Not running')
  }
})

ipc.on('start-server', (event) => {
  let proxy_ports = []
  // SSH tunnel localhost:3128 => docker0:3128
  if (squid_proxy === undefined ) {
    event.sender.send('starting-server', 'Starting HTTP(S) proxy tunnel...')
	squid_proxy = net.createServer(function(sock) {
	  if (debug) {
	    console.log(sock)
	  }
	  let conn = new Client()
	  conn.on('ready', function() {
		conn.forwardOut(sock.remoteAddress, sock.remotePort, proxy_remote, squid_port, function(err, stream) {
		  if (err) {
			conn.end()
			return deny()
		  }
		  if (debug) {
			console.log(stream)
		  }
		  stream.pipe(sock).on('error', function (err) {
		    deny()
		  }).on('close', function (err) {
		    conn.end();
		  }).pipe(stream).on('error', function (err) {
		    deny()
		  }).on('close', function() {
	  	    conn.end()
	  	  })
		})
	  }).on('error', function(err) {
		deny()
	  }).connect(ssh_config)
	}).listen(squid_port, '::', () => {
	  if (!proxy_ports.indexOf(squid_port) > -1) { proxy_ports.push(squid_port) }
	  message = 'Proxy tunnel(s) localhost: ' + proxy_ports
	  event.sender.send('started-server', message)
	})
  } else {
    event.sender.send('started-server', 'Already running')
  }

  // SSH tunnel localhost:3128 => docker0:3128
  if (socks_proxy === undefined ) {
    event.sender.send('starting-server', 'Starting SOCKS proxy tunnel...')
	socks_proxy = net.createServer(function(sock) {
	  if (debug) {
	    console.log(sock)
	  }
	  let conn = new Client()
	  conn.on('ready', function() {
		conn.forwardOut(sock.remoteAddress, sock.remotePort, proxy_remote, socks_port, function(err, stream) {
		  if (err) {
			conn.end()
			return deny()
		  }
		  if (debug) {
			console.log(stream)
		  }
		  stream.pipe(sock).on('error', function (err) {
		    deny()
		  }).on('close', function (err) {
		    conn.end();
		  }).pipe(stream).on('error', function (err) {
		    deny()
		  }).on('close', function() {
	  	    conn.end()
	  	  })
		})
	  }).on('error', function(err) {
		deny()
	  }).connect(ssh_config)
	}).listen(socks_port, '::', () => {
	  if (!proxy_ports.indexOf(socks_port) > -1) { proxy_ports.push(socks_port) }
	  message = 'Proxy tunnel(s) localhost: ' + proxy_ports
	  event.sender.send('started-server', message)
	})
  } else {
    event.sender.send('started-server', 'Already running')
  }
})

ipc.on('stop-server', (event) => {
  if (squid_proxy !== undefined ) {
    event.sender.send('stopping-server', 'Stopping tunnel...')
    squid_proxy.close()
    squid_proxy = undefined
    event.sender.send('stopped-server', 'Tunnel stopped')
  } else {
    event.sender.send('stopped-server', 'Not running')
  }

  if (socks_proxy !== undefined ) {
    event.sender.send('stopping-server', 'Stopping tunnel...')
    socks_proxy.close()
    socks_proxy = undefined
    event.sender.send('stopped-server', 'Tunnel stopped')
  } else {
    event.sender.send('stopped-server', 'Not running')
  }
})
